File: microservice/template-service-python/handlers.py
import logging
from models import StreamMessage, EchoParameter, AddParameter, SubtractParameter, MultiplyParameter, DivideParameter


class Handlers:
    def __init__(self, service):
        self.service = service
        self.logger = logging.getLogger(__name__)
    
    def handle_echo(self, msg: StreamMessage):
        """处理echo服务"""
        if self.service.debug:
            self.logger.info(f"Handling echo message: {msg.message_id}")
        
        try:
            # 解析参数
            params = self.service.parse_parameters(msg, EchoParameter)
            
            # 处理业务逻辑（echo返回相同的消息）
            self.service.send_success_response(msg, {
                "echo": params.message
            })
            
        except Exception as e:
            self.logger.error(f"Error handling echo: {e}")
            self.service.send_error_response(msg, f"Invalid parameters: {str(e)}")
    
    def handle_add(self, msg: StreamMessage):
        """处理加法服务"""
        if self.service.debug:
            self.logger.info(f"Handling add message: {msg.message_id}")
        
        try:
            # 解析参数
            params = self.service.parse_parameters(msg, AddParameter)
            
            # 处理业务逻辑（加法运算）
            result = params.num1 + params.num2
            self.service.send_success_response(msg, {
                "result": result,
                "num1": params.num1,
                "num2": params.num2
            })
            
        except Exception as e:
            self.logger.error(f"Error handling add: {e}")
            self.service.send_error_response(msg, f"Invalid parameters: {str(e)}")
    
    def handle_subtract(self, msg: StreamMessage):
        """处理减法服务"""
        if self.service.debug:
            self.logger.info(f"Handling subtract message: {msg.message_id}")
        
        try:
            # 解析参数
            params = self.service.parse_parameters(msg, SubtractParameter)
            
            # 处理业务逻辑（减法运算）
            result = params.num1 - params.num2
            self.service.send_success_response(msg, {
                "result": result,
                "num1": params.num1,
                "num2": params.num2
            })
            
        except Exception as e:
            self.logger.error(f"Error handling subtract: {e}")
            self.service.send_error_response(msg, f"Invalid parameters: {str(e)}")
    
    def handle_multiply(self, msg: StreamMessage):
        """处理乘法服务"""
        if self.service.debug:
            self.logger.info(f"Handling multiply message: {msg.message_id}")
        
        try:
            # 解析参数
            params = self.service.parse_parameters(msg, MultiplyParameter)
            
            # 处理业务逻辑（乘法运算）
            result = params.num1 * params.num2
            self.service.send_success_response(msg, {
                "result": result,
                "num1": params.num1,
                "num2": params.num2
            })
            
        except Exception as e:
            self.logger.error(f"Error handling multiply: {e}")
            self.service.send_error_response(msg, f"Invalid parameters: {str(e)}")
    
    def handle_divide(self, msg: StreamMessage):
        """处理除法服务"""
        if self.service.debug:
            self.logger.info(f"Handling divide message: {msg.message_id}")
        
        try:
            # 解析参数
            params = self.service.parse_parameters(msg, DivideParameter)
            
            # 检查除数是否为零
            if params.num2 == 0:
                self.service.send_error_response(msg, "Division by zero is not allowed")
                return
            
            # 处理业务逻辑（除法运算）
            result = params.num1 / params.num2
            self.service.send_success_response(msg, {
                "result": result,
                "num1": params.num1,
                "num2": params.num2
            })
            
        except Exception as e:
            self.logger.error(f"Error handling divide: {e}")
            self.service.send_error_response(msg, f"Invalid parameters: {str(e)}")

File: microservice/template-service/handlers.go
package main

import (
	"encoding/json"
	"log"
	"strconv"

	"WebSupervisor/MyTool/streamtool/models"
	"WebSupervisor/model"
)

// Example handlers for demonstration
func handleStreamMessage(){

	
}
// HandleEcho 处理echo服务
func (s *TemplateService) HandleEcho(msg *models.StreamMessage) {
	if s.debug {
		log.Printf("Handling echo message: %s", msg.MessageID)
	}

	// 解析参数
	var params model.EchoParameter
	if err := s.parseParameters(msg, &params); err != nil {
		s.sendErrorResponse(msg, "Invalid parameters: "+err.Error())
		return
	}

	// 处理业务逻辑（echo返回相同的消息）
	s.sendSuccessResponse(msg, map[string]interface{}{
		"echo": params.Message,
	})
}

// HandleAdd 处理加法服务
func (s *TemplateService) HandleAdd(msg *models.StreamMessage) {
	if s.debug {
		log.Printf("Handling add message: %s", msg.MessageID)
	}

	// 解析参数
	var params model.AddParameter
	if err := s.parseParameters(msg, &params); err != nil {
		s.sendErrorResponse(msg, "Invalid parameters: "+err.Error())
		return
	}

	// 处理业务逻辑（加法运算）
	result := params.Num1 + params.Num2
	s.sendSuccessResponse(msg, map[string]interface{}{
		"result": result,
		"num1":   params.Num1,
		"num2":   params.Num2,
	})
}

// parseParameters 解析消息参数
func (s *TemplateService) parseParameters(msg *models.StreamMessage, params interface{}) error {
	jsonData, err := json.Marshal(msg.Playload)
	if err != nil {
		return err
	}
	return json.Unmarshal(jsonData, params)
}

// sendResponse 发送响应消息
func (s *TemplateService) sendResponse(msg *models.StreamMessage, success bool, data map[string]interface{}) {
	responseMsg := &models.StreamMessage{
		MessageID:      strconv.Itoa(s.stream.GetMessageID()),
		ReplyID:        msg.MessageID,
		ServiceName:    "response",
		CallbackStream: msg.CallbackStream,
		Playload: map[string]interface{}{
			"success": success,
			"data":    data,
		},
	}

	s.stream.StreamPush(responseMsg, msg.CallbackStream)
}

// sendSuccessResponse 发送成功响应
func (s *TemplateService) sendSuccessResponse(msg *models.StreamMessage, data map[string]interface{}) {
	s.sendResponse(msg, true, data)
}

// sendErrorResponse 发送错误响应
func (s *TemplateService) sendErrorResponse(msg *models.StreamMessage, errorMsg string) {
	s.sendResponse(msg, false, map[string]interface{}{
		"error": errorMsg,
	})
}